introspection-engine/connectors/sql-introspection-connector/src/re_introspection.rs

```diff
@@ -135,61 +137,40 @@ pub fn enrich(old_data_model: &Datamodel, new_data_model: &mut Datamodel) -> Vec
-    // todo this only updates relation names where the name of a model changed.
-    // the change of a field name would here not be reflected yet, but these would have been rendered already
+    //keep old virtual relation names on non M:N relations
+    // M:N relations cannot be uniquely identified without ignoring the relationname and their relationnames cant
+    // be changed without necessitation db changes since RelationName -> Join table name
+    let mut changed_relation_names = vec![];
     {
-        let mut relation_fields_to_change = vec![];
-        for changed_model_name in &changed_model_names {
-            let changed_model = new_data_model.find_model(&changed_model_name.1.model).unwrap();
-
-            for rf in changed_model.relation_fields() {
-                let info = &rf.relation_info;
-                let other_model_in_relation = new_data_model.find_model(&info.to).unwrap();
-                let number_of_relations_to_other_model_in_relation = changed_model
-                    .relation_fields()
-                    .filter(|f| f.points_to_model(&info.to))
-                    .count();
-
-                let other_relation_field = new_data_model.find_related_field_bang(&rf);
-                let other_info = &other_relation_field.relation_info;
-
-                let (model_with_fk, referenced_model, fk_column_name) = if info.to_fields.is_empty() {
-                    // does not hold the fk
-                    (
-                        &other_model_in_relation.name,
-                        &changed_model.name,
-                        other_info.fields.join("_"),
-                    )
-                } else {
-                    // holds the fk
-                    (
-                        &changed_model.name,
-                        &other_model_in_relation.name,
-                        info.fields.join("_"),
-                    )
-                };
-
-                let unambiguous = number_of_relations_to_other_model_in_relation < 2;
-                let relation_name = if unambiguous {
-                    RelationNames::name_for_unambiguous_relation(model_with_fk, referenced_model)
-                } else {
-                    RelationNames::name_for_ambiguous_relation(model_with_fk, referenced_model, &fk_column_name)
-                };
-
-                relation_fields_to_change.push((changed_model.name.clone(), rf.name.clone(), relation_name.clone()));
-                relation_fields_to_change.push((
-                    other_model_in_relation.name.clone(),
-                    other_relation_field.name.clone(),
-                    relation_name.clone(),
-                ));
+        for model in new_data_model.models() {
+            for field in model.relation_fields() {
+                if let Some(old_model) = old_data_model.find_model(&model.name) {
+                    for old_field in old_model.relation_fields() {
```

Comment:
These nested loops and conditionals are making me think we could structure reintrospection a bit more like the sql schema differ in the migration engine. For example here, there are 4 level of nestings for an action that is conceptually pretty simple: comparing the relation fields for models that are present in both the new and the old schemas. Maybe we could build to an API like `reintrospector.model_pairs().flat_map(|models| models.relation_field_pairs()).for_each(|relation_fields| { /* logic goes here */ })` — in the future.

Reply:
(we can even add a `.filter_map(|relation_fields| relation_fields.as_m2m())` and then be sure we are dealing only with many-to-many relations)

Reply:
Yeah, let's pair on that once we have a bit of time.